Mathlib/Algebra/Lie/Rank.lean

```diff
@@ -0,0 +1,251 @@
+/-
+Copyright (c) 2024 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin
+-/
+import Mathlib.Algebra.Lie.OfAssociative
+import Mathlib.Data.MvPolynomial.Monad
+import Mathlib.LinearAlgebra.Charpoly.ToMatrix
+import Mathlib.LinearAlgebra.Dimension.Finrank
+import Mathlib.LinearAlgebra.FreeModule.StrongRankCondition
+import Mathlib.LinearAlgebra.Matrix.Charpoly.Univ
+
+/-!
+# Rank of a Lie algebra and regular elements
+
+Let `L` be a Lie algebra over a nontrivial commutative ring `R`,
+and assume that `L` is finite free as `R`-module.
+Then the coefficients of the characteristic polynomial of `ad R L x` are polynomial in `x`.
+The *rank* of `L` is the smallest `n` for which the `n`-th coefficient is not the zero polynomial.
+
+Continuing to write `n` for the rank of `L`, an element `x` of `L` is *regular*
+if the `n`-th coefficient of the characteristic polynomial of `ad R L x` is non-zero.
+
+## Main declarations
+
+* `LieAlgebra.rank R L` is the rank of a Lie algebra `L` over a commutative ring `R`.
+* `LieAlgebra.IsRegular R x` is the predicate that an element `x` of a Lie algebra `L` is regular.
+
+## References
+
+* [barnes1967]: "On Cartan subalgebras of Lie algebras" by D.W. Barnes.
+
+-/
+
+open scoped BigOperators
+
+variable {R L M ι ι' ιM ιM' : Type*}
+variable [CommRing R] [LieRing L] [LieAlgebra R L]
+variable [AddCommGroup M] [Module R M] [LieRingModule L M] [LieModule R L M]
+variable [Fintype ι] [Fintype ιM] [Fintype ι'] [Fintype ιM']
+
+namespace LieAlgebra
+
+local notation "φ" => LieModule.toEndomorphism R L M
+
+section basic
+
+variable [DecidableEq ιM] [DecidableEq ιM']
+variable (b : Basis ι R L) (bₘ : Basis ιM R M) (b' : Basis ι' R L) (bₘ' : Basis ιM' R M)
+
+open LieModule LinearMap MvPolynomial in
+/-- Let `M` be a Lie module of a Lie algebra `L` over `R`,
+and let `b` be a basis of `L` and `bₘ` a basis of `M`.
+Then `lieMatrixPoly b bₘ (i,j)` is the polynomial that evaluates on elements `x` of `L`
+to the `(i,j)`-th coefficient of the matrix representation of `⁅x, ·⁆` acting on `M`. -/
+noncomputable
+def lieMatrixPoly (ij : ιM × ιM) : MvPolynomial ι R :=
+  ∑ k : ι, monomial (.single k 1) ((toMatrix bₘ bₘ <| φ <| b k) ij.1 ij.2)
```

Comment:
I guess if we didn't have the constraint that Mathlib currently only supports polynomials with commutative coefficients, this could be:
```lean
def lieMatrixPoly' : MvPolynomial ι (Module.End R M) :=
   ∑ k : ι, monomial (.single k 1) (φ <| b k)
```
It might be worth adding a sentence to the doc string saying that this definition is morally a single-variable polynomial with coefficients in `Module.End R M`.

Reply:
I think noncommutative coefficients are already supported (or maybe only for `Polynomial`?)... but I really want something with coefficients in `R`, I think. It's not clear to me how I can use your proposed definition to extract the fact that the characteristic polynomial of `ad x` has coefficients that are "polynomial in `x`".

Reply:
Yes, non-commutatve coefficients are permitted for `Polynomial` but not for `MvPolynomial`. Do you think it might be worth adding a sentence in the doc string about this being morally a single-variable polynomial with coefficients in `Module.End R M`.

Reply:
This whole theory has been factored out into a new file. So this is resolved.